xs/src/libslic3r/ConditionalGcode.cpp

```diff
@@ -0,0 +1,133 @@
+#include <string>
+
+#include <sstream>
+#include <exprtk/exprtk.hpp>
+#include "ConditionalGcode.hpp"
+namespace Slic3r {
+
+// trim from start (in place)
+static inline void ltrim(std::string &s) {
+    s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](int ch) {
+        return !std::isspace(ch);
+    }));
+}
+
+// trim from end (in place)
+static inline void rtrim(std::string &s) {
+    s.erase(std::find_if(s.rbegin(), s.rend(), [](int ch) {
+        return !std::isspace(ch);
+    }).base(), s.end());
+}
+
+// trim from both ends (in place)
+static inline void trim(std::string &s) {
+    ltrim(s);
+    rtrim(s);
+}
+/// Start of recursive function to parse gcode file.
+std::string apply_math(const std::string& input) {
+    return expression(input);
+}
+
+/// Evaluate expressions with exprtk
+/// Everything must resolve to a number.
+std::string evaluate(const std::string& expression_string) {
+    std::stringstream result;
+
+    #if SLIC3R_DEBUG
+    std::cerr << __FILE__ << ":" << __LINE__ << " "<< "Evaluating expression: " << expression_string << std::endl;
+    #endif
+    double num_result = double(0);
+    if ( exprtk::compute(expression_string, num_result)) { 
+        result << num_result;
+    } else {
+        #if SLIC3R_DEBUG
+        std::cerr << __FILE__ << ":" << __LINE__ << " "<< "Failed to parse: " << expression_string.c_str() << std::endl;
+        #endif
+    }
+    std::string output = result.str();
+    trim(output);
+    return output;
+} 
+
+
+/// Parse an expression and return a string. We assume that PlaceholderParser has expanded all variables. 
+std::string expression(const std::string& input, const int depth) {
```

Comment:
With `exprtk` all that is IMO necessary is to replace (multiple) `{expr}` with its evaluated form (ideally including string result) and handle `{if expr}` as conditional GCode. There is no need to handle subexpressions at this level (and it will interfere with exprtk)

With string support, there will appear problem with curly brackets embedded in strings. 
Minor problem may be with embedded placeholders, but it's easy to workaround (split text to avoid placeholder expansion)